File: backend-java/src/test/java/com/bffgin/backend/test_support/MockJwksServer.java
package com.bffgin.backend.test_support;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.security.interfaces.RSAPublicKey;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

/**
 * テスト専用の自プロセス内蔵JWKSサーバー(backend-c/tests/jwks_test.c・
 * backend-cpp/tests/mock_jwks_server.{hpp,cpp}と同じ役割)。
 * 実際にKeycloak/bffを起動せずにJwksVerifierのRS256検証・kidキャッシュ・
 * 未知kid時の再取得ロジックを検証できる
 */
public final class MockJwksServer implements AutoCloseable {

    private final HttpServer server;
    private final List<RSAPublicKey> keys = new ArrayList<>();
    private final List<String> kids = new ArrayList<>();
    private final ObjectMapper mapper = new ObjectMapper();

    public MockJwksServer() throws IOException {
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/jwks", exchange -> {
            byte[] body = buildJwksJson().getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, body.length);
            try (var os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.setExecutor(Executors.newVirtualThreadPerTaskExecutor());
        server.start();
    }

    public void addKey(String kid, RSAPublicKey key) {
        kids.add(kid);
        keys.add(key);
    }

    public String jwksUrl() {
        return "http://127.0.0.1:" + server.getAddress().getPort() + "/jwks";
    }

    private String buildJwksJson() {
        ObjectNode root = mapper.createObjectNode();
        ArrayNode keysNode = root.putArray("keys");
        for (int i = 0; i < keys.size(); i++) {
            RSAPublicKey key = keys.get(i);
            ObjectNode k = mapper.createObjectNode();
            k.put("kty", "RSA");
            k.put("kid", kids.get(i));
            k.put("use", "sig");
            k.put("n", TestTokenHelper.rsaModulusB64Url(key));
            k.put("e", TestTokenHelper.rsaExponentB64Url(key));
            keysNode.add(k);
        }
        return root.toString();
    }

    @Override
    public void close() {
        server.stop(0);
    }
}

File: backend-kotlin/src/test/kotlin/com/bffgin/backend/test_support/MockJwksServer.kt
package com.bffgin.backend.test_support

import com.fasterxml.jackson.databind.ObjectMapper
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import java.security.interfaces.RSAPublicKey
import java.util.concurrent.Executors

/**
 * テスト専用の自プロセス内蔵JWKSサーバー(backend-java/backend-c/backend-cppのモックJWKSサーバーと
 * 同じ役割)。実際にKeycloak/bffを起動せずにJwksVerifierのRS256検証・kidキャッシュ・
 * 未知kid時の再取得ロジックを検証できる。テスト専用の道具としてJDK標準のHttpServerを使う
 * (本番のREST実装はKtorだが、これはあくまでテスト用の使い捨てモックであり、
 * 「本番はKtor、Verifierが叩く先はただのHTTPサーバーであれば何でもよい」という点を示す)
 */
class MockJwksServer : AutoCloseable {

    private val server: HttpServer = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
    private val keys = mutableListOf<RSAPublicKey>()
    private val kids = mutableListOf<String>()
    private val mapper = ObjectMapper()

    init {
        server.createContext("/jwks") { exchange ->
            val body = buildJwksJson().toByteArray(StandardCharsets.UTF_8)
            exchange.responseHeaders.add("Content-Type", "application/json")
            exchange.sendResponseHeaders(200, body.size.toLong())
            exchange.responseBody.use { it.write(body) }
        }
        server.executor = Executors.newVirtualThreadPerTaskExecutor()
        server.start()
    }

    fun addKey(kid: String, key: RSAPublicKey) {
        kids.add(kid)
        keys.add(key)
    }

    fun jwksUrl(): String = "http://127.0.0.1:${server.address.port}/jwks"

    private fun buildJwksJson(): String {
        val root = mapper.createObjectNode()
        val keysNode = root.putArray("keys")
        for (i in keys.indices) {
            val key = keys[i]
            val k = mapper.createObjectNode()
            k.put("kty", "RSA")
            k.put("kid", kids[i])
            k.put("use", "sig")
            k.put("n", TestTokenHelper.rsaModulusB64Url(key))
            k.put("e", TestTokenHelper.rsaExponentB64Url(key))
            keysNode.add(k)
        }
        return root.toString()
    }

    override fun close() {
        server.stop(0)
    }
}
